internal/resources/gsi.go

```diff
@@ -0,0 +1,509 @@
+		state.ClusterId = types.StringValue(clusterID)
+		state.BucketName = types.StringValue(bucketName)
+		state.ScopeName = types.StringValue(scopeName)
+		state.CollectionName = types.StringValue(collectionName)
+		state.IsPrimary = types.BoolValue(index.IsPrimary)
+		state.IndexName = types.StringValue(indexName)
+
+		var keys []attr.Value
+		for _, key := range index.SecExprs {
+			keys = append(keys, types.StringValue(key))
+		}
+
+		keyList, newDiags := types.ListValue(types.StringType, keys)
+		if newDiags.HasError() {
+			resp.Diagnostics.AddError(
+				"Error converting index keys to set type",
+				"Could not convert index keys to set type for index "+state.IndexName.ValueString()+": "+err.Error(),
+			)
+			return
+		}
+
+		state.IndexKeys = keyList
+		// TODO:  set partition by.
+		state.Where = types.StringValue(index.Where)
+		if state.With != nil {
+			state.With.NumReplica = types.Int64Value(int64(index.NumReplica))
+			state.With.NumPartition = types.Int64Value(int64(index.NumPartition))
+		}
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+func (g *GSI) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var plan providerschema.GsiDefinition
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	ddl := fmt.Sprintf(
+		"ALTER INDEX `%s` ON `%s`.`%s`.`%s` WITH { \"action\": \"replica_count\", \"num_replica\" : %d }",
+		plan.IndexName.ValueString(),
+		plan.BucketName.ValueString(),
+		plan.ScopeName.ValueString(),
+		plan.CollectionName.ValueString(),
+		plan.With.NumReplica.ValueInt64(),
+	)
+
+	if err := g.executeGsiDdl(ctx, &plan, ddl); err != nil {
+		resp.Diagnostics.AddError(
+			"An error occurred while executing index DDL",
+			"Error during index DDL execution: "+err.Error(),
+		)
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+func (g *GSI) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	var state providerschema.GsiDefinition
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if !state.BuildIndexes.IsNull() {
+		return
+	}
+
+	var indexName string
+	if state.IsPrimary.ValueBool() && state.IndexName.IsNull() {
+		indexName = "#primary"
+	} else {
+		indexName = state.IndexName.ValueString()
+	}
+
+	ddl := fmt.Sprintf(
+		"DROP INDEX `%s` ON `%s`.`%s`.`%s`",
+		indexName,
+		state.BucketName.ValueString(),
+		state.ScopeName.ValueString(),
+		state.CollectionName.ValueString(),
+	)
+
+	if err := g.executeGsiDdl(ctx, &state, ddl); err != nil {
+		resp.Diagnostics.AddError(
+			"An error occurred while executing index DDL",
+			"Error during index DDL execution: "+err.Error(),
+		)
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, state)...)
```

Comment:
Is there any specific reason why we are saving back to the state file during the delete operation when no error is encountered?

Reply:
Good catch.  Should not save to state file.

Addressed in 218633a9d35cfe509d718b3d1994beb6b0162120